README.md
# GoogleMapsGeocoder

[![Build Status](https://secure.travis-ci.org/trystant/google_maps_geocoder.png)](http://travis-ci.org/trystant/google_maps_geocoder)
[![Maintainability](https://api.codeclimate.com/v1/badges/128cc62114c6575fbcb2/maintainability)](https://codeclimate.com/github/trystant/google_maps_geocoder/maintainability)
[![Coverage Status](https://coveralls.io/repos/github/trystant/google_maps_geocoder/badge.svg?branch=master)](https://coveralls.io/github/trystant/google_maps_geocoder?branch=master)
[![Dependency Status](https://gemnasium.com/trystant/google_maps_geocoder.png)](https://gemnasium.com/trystant/google_maps_geocoder)
[![Inline docs](http://inch-ci.org/github/trystant/google_maps_geocoder.svg?branch=master)](http://inch-ci.org/github/trystant/google_maps_geocoder)
[![Gem Version](https://badge.fury.io/rb/google_maps_geocoder.svg)](http://badge.fury.io/rb/google_maps_geocoder)

A simple Plain Old Ruby Object wrapper for geocoding with Google Maps.

## Installation

Add GoogleMapsGeocoder to your Gemfile and run `bundle`:

```ruby
  gem 'google_maps_geocoder'
```

Or try it out in `irb` with:

```ruby
  require './lib/google_maps_geocoder/google_maps_geocoder'
```

## Ready to Go in One Step

```ruby
chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania Washington'
```

## Usage

Get the complete, formatted address:

```ruby
chez_barack.formatted_address
 => "1600 Pennsylvania Avenue Northwest, President's Park, Washington, DC 20500, USA"
```

...standardized name of the city:

```ruby
chez_barack.city
 => "Washington"
```

...full name of the state or region:

```ruby
chez_barack.state_long_name
 => "District of Columbia"
```

...standard abbreviation for the state/region:

```ruby
chez_barack.state_short_name
 => "DC"
```

## API

The complete, hopefully self-explanatory, API is:

* `GoogleMapsGeocoder#city`
* `GoogleMapsGeocoder#country_long_name`
* `GoogleMapsGeocoder#country_short_name`
* `GoogleMapsGeocoder#county`
* `GoogleMapsGeocoder#exact_match?`
* `GoogleMapsGeocoder#formatted_address`
* `GoogleMapsGeocoder#formatted_street_address`
* `GoogleMapsGeocoder#lat`
* `GoogleMapsGeocoder#lng`
* `GoogleMapsGeocoder#partial_match?`
* `GoogleMapsGeocoder#postal_code`
* `GoogleMapsGeocoder#state_long_name`
* `GoogleMapsGeocoder#state_short_name`

## Google Maps API Key (Optional)

To have GoogleMapsGeocoder use your Google Maps API key, set it as an environment variable:

```bash
export GOOGLE_MAPS_API_KEY=[your key]
```

## Contributing to google_maps_geocoder

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Copyright

Copyright © 2011-2017 Roderick Monje. See LICENSE.txt for further details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 0.5.1 - 2018-01-22
### Changed
- Updated Ruby to 2.5.0

spec/spec_helper.rb
require 'simplecov'
SimpleCov.start
require 'coveralls'
Coveralls.wear!
require 'dotenv'
Dotenv.load
require 'pry'
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  Coveralls::SimpleCov::Formatter
]
require 'rubygems'
require 'bundler'
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  warn e.message
  warn 'Run `bundle install` to install missing gems'
  exit e.status_code
end
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'google_maps_geocoder/google_maps_geocoder'


spec/lib/google_maps_geocoder_spec.rb
require File.dirname(__FILE__) + '/../spec_helper'
# rubocop:disable Metrics/BlockLength
describe GoogleMapsGeocoder do
  before(:all) do
    begin
      @exact_match   = GoogleMapsGeocoder.new('837 Union Street Brooklyn NY')
      @partial_match = GoogleMapsGeocoder.new('1600 Pennsylvania Washington')
    rescue SocketError
      @no_network  = true
    rescue RuntimeError
      @query_limit = true
    end
  end
  # rubocop:enable Metrics/BlockLength
  before(:each) do
    pending 'waiting for a network connection' if @no_network
    pending 'waiting for query limit to pass' if @query_limit
  end

  context 'with "837 Union Street Brooklyn NY"' do
    subject { @exact_match }

    it { expect(subject).to be_exact_match }

    context 'address' do
      it { expect(subject.formatted_street_address).to eq '837 Union Street' }
      it { expect(subject.city).to eq 'Brooklyn' }
      it { expect(subject.county).to match(/Kings/) }
      it { expect(subject.state_long_name).to eq 'New York' }
      it { expect(subject.state_short_name).to eq 'NY' }
      it { expect(subject.postal_code).to match(/112[0-9]{2}/) }
      it { expect(subject.country_short_name).to eq 'US' }
      it { expect(subject.country_long_name).to eq 'United States' }
      it do
        expect(subject.formatted_address)
          .to match(/837 Union St, Brooklyn, NY 112[0-9]{2}, USA/)
      end
    end
    context 'coordinates' do
      it { expect(subject.lat).to be_within(0.005).of(40.6748151) }
      it { expect(subject.lng).to be_within(0.005).of(-73.9760302) }
    end
  end

  context 'with "1600 Pennsylvania Washington"' do
    subject { @partial_match }

    it { expect(subject).to be_partial_match }

    context 'address' do
      it do
        expect(subject.formatted_street_address)
          .to match '1600 Pennsylvania Avenue'
      end
      it { expect(subject.city).to eq 'Washington' }
      it { expect(subject.state_long_name).to eql 'District of Columbia' }
      it { expect(subject.state_short_name).to eql 'DC' }
      it { expect(subject.country_short_name).to eql 'US' }
      it { expect(subject.country_long_name).to eql 'United States' }
      it do
        expect(subject.formatted_address)
          .to match(/1600 Pennsylvania Ave.*, Washington, DC.*, USA/)
      end
    end

    context 'coordinates' do
      it { expect(subject.lat).to be_within(0.5).of(38.8791708) }
      it { expect(subject.lng).to be_within(0.5).of(-76.9818148) }
    end
  end

  context "when ENV['GOOGLE_MAPS_API_KEY'] is invalid" do
    subject { @exact_match }

    it do
      allow(subject).to receive(:send).with(:query_url, nil).and_return(
        'https://maps.googleapis.com/maps/api/geocode/json?address='\
        '&sensor=false&key=INVALID_KEY'
      )
      expect(subject.send(:query_url, nil)).to eql(
        'https://maps.googleapis.com/maps/api/geocode/json?address='\
        '&sensor=false&key=INVALID_KEY'
      )
    end
  end

  context 'with google returns empty results' do
    let(:results_hash) { { 'results' => [] } }

    GoogleMapsGeocoder::ERROR_STATUSES.each do |key, value|
      it "raises #{key} error" do
        allow_any_instance_of(GoogleMapsGeocoder).to receive(:json_from_url)
          .and_return results_hash.merge('status' => value)
        expect { GoogleMapsGeocoder.new('anything') }
          .to raise_error(GoogleMapsGeocoder.send(:error_class_name, key))
      end
    end
  end
end
# rubocop:disable Metrics/BlockLength
describe GoogleMapsGeocoder, 'batch' do
  before(:all) do
    begin
      address_hash = {
        '1': '837 Union Street Brooklyn NY',
        '2': '1600 Pennsylvania Washington'
      }
      @array_match = GoogleMapsGeocoder.new(address_hash)
      p 'Done with array constructor'
    rescue SocketError
      @no_network  = true
    rescue RuntimeError
      @query_limit = true
    rescue StandardError
      @unknown_error = true
    end
  end

  # rubocop:enable Metrics/BlockLength
  before(:each) do
    pending 'waiting for a network connection' if @no_network
    pending 'waiting for query limit to pass' if @query_limit
    pending 'an unknown error occurred' if @unknown_error
  end

  context 'with multiple addresses' do
    pending 'not ready yet' do
      it { expect(subject).to be_array_match }
    end
  end
end

describe GoogleMapsGeocoder, '#build_google_api_urls' do
  let(:uri) { GoogleMapsGeocoder.api_uri }
  let(:address_hash) do
    {
      1 => '837 Union Street Brooklyn NY',
      2 => '1600 Pennsylvania Washington'
    }
  end

  it 'should turn a hash with IDs & addresses into a hash with IDs and URLs' do
    geocoder = GoogleMapsGeocoder.new(nil)
    address_with_urls_hash = geocoder.build_google_api_urls(address_hash)
    expect(address_with_urls_hash).to eq(
      1 => "#{uri}?address=837+Union+Street+Brooklyn+NY&sensor=false&"\
        "key=#{ENV['GOOGLE_MAPS_API_KEY']}",
      2 => "#{uri}?address=1600+Pennsylvania+Washington&sensor=false&"\
        "key=#{ENV['GOOGLE_MAPS_API_KEY']}"
    )
  end
end

describe GoogleMapsGeocoder, '#make_requests' do
  let(:uri) { GoogleMapsGeocoder.api_uri }
  let(:google_api_urls) do
    {
      1 => "#{uri}?address=837+Union+Street+Brooklyn+NY&sensor=false&"\
        "key=#{ENV['GOOGLE_MAPS_API_KEY']}",
      2 => "#{uri}?address=1600+Pennsylvania+Washington&sensor=false&"\
        "key=#{ENV['GOOGLE_MAPS_API_KEY']}"
    }
  end
  it 'returns a hash of GoogleMapsGeocoder results given a hash with URLs'
end

describe GoogleMapsGeocoder, '.api_uri' do
  it 'returns the URL for the API' do
    expect(GoogleMapsGeocoder.api_uri).to eq('https://maps.googleapis.com/maps/api/geocode/json')
  end
end


lib/google_maps_geocoder/google_maps_geocoder.rb
require 'active_support'
require 'active_support/core_ext/string/inflections'
require 'logger'
require 'net/http'
require 'rack'
require 'curb'
# A simple PORO wrapper for geocoding with Google Maps.
#
# @example
#   chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania Ave'
#   chez_barack.formatted_address
#     => "1600 Pennsylvania Avenue Northwest, President's Park,
#         Washington, DC 20500, USA"
# rubocop:disable Metrics/ClassLength
# rubocop:disable Metrics/AbcSize
class GoogleMapsGeocoder
  class << self
    # returns the Google Map API URL
    def api_uri
      GOOGLE_API_URI
    end
  end
  # Error handling for google statuses
  class GeocodingError < StandardError
    # Initialize an error class wrapping the error returned by Google Maps.
    #
    # @return [GeocodingError] the geocoding error
    def initialize(response_json = '')
      @json = response_json
      super
    end

    # Returns the GeocodingError's content.
    #
    # @return [String] the geocoding error's content
    def message
      "Google returned:\n#{@json.inspect}"
    end
  end

  class ZeroResultsError < GeocodingError; end
  class QueryLimitError < GeocodingError; end
  class RequestDeniedError < GeocodingError; end
  class InvalidRequestError < GeocodingError; end
  class UnknownError < GeocodingError; end

  ERROR_STATUSES = { zero_results: 'ZERO_RESULTS',
                     query_limit: 'OVER_QUERY_LIMIT',
                     request_denied: 'REQUEST_DENIED',
                     invalid_request: 'INVALID_REQUEST',
                     unknown: 'UNKNOWN_ERROR' }.freeze

  GOOGLE_ADDRESS_SEGMENTS = %i[
    city
    country_long_name
    country_short_name
    county
    lat
    lng
    neighborhood
    bounds
    postal_code
    state_long_name
    state_short_name
  ].freeze
  GOOGLE_API_URI = 'https://maps.googleapis.com/maps/api/geocode/json'.freeze

  ALL_ADDRESS_SEGMENTS = (
    GOOGLE_ADDRESS_SEGMENTS + %i[
      formatted_address formatted_street_address
    ]
  ).freeze

  # Returns the complete formatted address with standardized abbreviations.
  #
  # @return [String] the complete formatted address
  # @example
  #   chez_barack.formatted_address
  #     => "1600 Pennsylvania Avenue Northwest, President's Park,
  #         Washington, DC 20500, USA"
  attr_reader :formatted_address

  # Returns the formatted street address with standardized abbreviations.
  #
  # @return [String] the formatted street address
  # @example
  #   chez_barack.formatted_street_address
  #     => "1600 Pennsylvania Avenue"
  attr_reader :formatted_street_address
  # Self-explanatory
  attr_reader(*GOOGLE_ADDRESS_SEGMENTS)

  # Returns a list of geocoded addresses
  def bulk_addresses
    @addresses
  end

  # Geocodes the specified address and wraps the results in a GoogleMapsGeocoder
  # object.
  #
  # @param data [String] a geocodable address
  # @return [GoogleMapsGeocoder] the Google Maps result for the specified
  #   address
  # @example
  #   chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania Ave'

  def build_google_api_urls(uris)
    urls = {}
    uris.each_key do |id|
      urls[id.to_i] = URI.parse(query_url(uris[id])).to_s
    end
    urls
  end

  def make_requests(urls) # rubocop:disable Metrics/MethodLength
    results = {}, easy_options = { follow_location: true }
    multi_options = { pipeline: Curl::CURLPIPE_MULTIPLEX } unless ENV['CI']
    Curl::Multi.get(urls.values, easy_options, multi_options) do |easy|
      begin
        results[urls.key(easy.last_effective_url)] =
          ActiveSupport::JSON.decode(easy.body_str)
      rescue StandardError => error
        p "error: #{error}"
      end
    end
    results
  end
  # rubocop:enable Metrics/MethodLength

  def initialize(data)
    initialize_single_address(data) if data.is_a?(String)
    initialize_multiple_addresses(data) if data.is_a?(Hash)
  end

  # initialization for single address
  def initialize_single_address(data)
    @json = data.is_a?(String) ? json_from_url(data) : data
    handle_error if @json.blank? || @json['status'] != 'OK'
    set_attributes_from_json
    logger.info('GoogleMapsGeocoder') do
      "Geocoded \"#{data}\" => \"#{formatted_address}\""
    end
  end

  # initialization for multiple addresses
  # rubocop:disable Metrics/MethodLength
  def initialize_multiple_addresses(data)
    @addresses = {}
    google_api_urls = build_google_api_urls(data)
    json_results = make_requests(google_api_urls)
    json_results.each do |key|
      next if key.empty?
      @json = key['results']
      bulk_attributes_from_json_for(@json)
      @addresses[id] = @json
      @json = nil
    end
    @addresses
  end
  # rubocop:enable Metrics/MethodLength

  # Fetches the neighborhood
  def fetch_neighborhood
    return unless bounds.is_a?(Array) && bounds.size == 4
    uri = URI.parse neighborhood_url
    logger.debug('GoogleMapsGeocoder') { uri }
    response = http(uri)
    results = ActiveSupport::JSON.decode response.body
    results['results'].map { |e| e['address_components'] }
                      .compact
                      .flatten
                      .select { |e| e['types'].include?('neighborhood') }
                      .map { |e| e['long_name'] }.uniq
  end

  # Bulk fetches the neighborhoods
  def fetch_neighborhoods
    return if @addresses.empty?
    uri = URI.parse neighborhood_url
    logger.debug('GoogleMapsGeocoder') { uri }
    response = http(uri)
    results = ActiveSupport::JSON.decode response.body
    results['results'].map { |e| e['address_components'] }
                      .compact
                      .flatten
                      .select { |e| e['types'].include?('neighborhood') }
                      .map { |e| e['long_name'] }.uniq
  end

  # Returns true if the address Google returns is an exact match.
  #
  # @return [boolean] whether the Google Maps result is an exact match
  # @example
  #   chez_barack.exact_match?
  #     => true
  def exact_match?
    !partial_match?
  end

  # Returns true if the address Google returns isn't an exact match.
  #
  # @return [boolean] whether the Google Maps result is a partial match
  # @example
  #   GoogleMapsGeocoder.new('1600 Pennsylvania Washington').partial_match?
  #     => true
  def partial_match?
    @json['results'][0]['partial_match'] == true
  end

  private

  def self.error_class_name(key)
    "google_maps_geocoder/#{key}_error".classify.constantize
  end
  private_class_method :error_class_name

  def api_key
    @api_key ||= "&key=#{ENV['GOOGLE_MAPS_API_KEY']}" if
      ENV['GOOGLE_MAPS_API_KEY']
  end

  def http(uri)
    c = Curl::Easy.new(uri.to_s) do |curl|
      curl.ssl_verify_peer = false
      curl.verbose = false
    end
    c.perform
    c
  end

  def json_from_url(url)
    uri = URI.parse query_url(url)
    logger.debug('GoogleMapsGeocoder') { uri }
    response = http(uri)
    ActiveSupport::JSON.decode response.body_str
  end

  def handle_error
    status = @json['status']
    message = GeocodingError.new(@json).message
    # for status codes see https://developers.google.com/maps/documentation/geocoding/intro#StatusCodes
    ERROR_STATUSES.each do |key, value|
      next unless status == value
      raise GoogleMapsGeocoder.send(:error_class_name, key), message
    end
  end

  def logger
    @logger ||= Logger.new STDERR
  end

  def parse_address_component_type(type, name = 'long_name')
    address_component = @json['results'][0]['address_components'].detect do |ac|
      ac['types'] && ac['types'].include?(type)
    end
    address_component && address_component[name]
  end

  def parse_city
    parse_address_component_type('sublocality') ||
      parse_address_component_type('locality')
  end

  def parse_country_long_name
    parse_address_component_type('country')
  end

  def parse_country_short_name
    parse_address_component_type('country', 'short_name')
  end

  def parse_county
    parse_address_component_type('administrative_area_level_2')
  end

  def parse_bounds
    northeast =  @json['results'][0]['geometry']['viewport']['northeast']
    southwest =  @json['results'][0]['geometry']['viewport']['southwest']
    [
      northeast['lat'],
      northeast['lng'],
      southwest['lat'],
      southwest['lng']
    ]
  end

  def parse_formatted_address
    @json['results'][0]['formatted_address']
  end

  def parse_formatted_street_address
    "#{parse_address_component_type('street_number')} "\
    "#{parse_address_component_type('route')}"
  end

  def parse_lat
    @json['results'][0]['geometry']['location']['lat']
  end

  def parse_lng
    @json['results'][0]['geometry']['location']['lng']
  end

  def parse_postal_code
    parse_address_component_type('postal_code')
  end

  def parse_neighborhood
    parse_address_component_type('neighborhood')
  end

  def parse_state_long_name
    parse_address_component_type('administrative_area_level_1')
  end

  def parse_state_short_name
    parse_address_component_type('administrative_area_level_1', 'short_name')
  end

  def query_url(query)
    "#{GOOGLE_API_URI}?address=#{Rack::Utils.escape query}&sensor=false"\
    "#{api_key}"
  end

  def neighborhood_url
    "#{GOOGLE_API_URI}?sensor=false"\
    "&latlng=#{@lat},#{@lng}"\
    '&componentRestrictions=neighborhood'\
    "#{api_key}"
  end

  def bulk_attributes_from_json_for(object)
    ALL_ADDRESS_SEGMENTS.each do |segment|
      begin
        object['results'][0][segment.to_s] = send("parse_#{segment}")
      rescue StandardError => error
        p "Error #{error}"
      end
    end
  end

  def set_attributes_from_json
    ALL_ADDRESS_SEGMENTS.each do |segment|
      instance_variable_set :"@#{segment}", send("parse_#{segment}")
    end
  end
end

# rubocop:enable Metrics/ClassLength
# rubocop:enable Metrics/AbcSize
